lib/route_consistency_checker.rb

```diff
@@ -0,0 +1,135 @@
+require "set"
+require "gds_api/router"
+
+class RouteConsistencyChecker
+  attr_reader :errors
+
+  def initialize(routes, router_data)
+    @routes = routes
+    @router_data = router_data
+    @checked_routes = Set.new
+    @errors = Hash.new { |hash, key| hash[key] = [] }
+  end
+
+  def check_content
+    content_items_to_check.each do |content_item|
+      content_item.redirects.each do |redirect|
+        check_redirect(content_item, redirect)
+      end
+
+      content_item.routes.each do |route|
+        check_route(content_item, route)
+      end
+    end
+  end
+
+  def check_routes
+    remaining_routes.each do |path|
+      next if router_data.include?(path)
+
+      route = routes.fetch(path)
+      next if route.disabled
+      next if route.handler == "gone"
+      next if route.handler == "redirect" && !route.backend_id.present?
+
+      content_item = find_content_item(route.incoming_path)
+      next if content_item
+
+      errors[route.incoming_path] << "No content item available."
+    end
+  end
+
+private
+
+  attr_reader :routes, :router_data, :checked_routes
+  attr_writer :errors
+
+  def content_items_to_check
+    ContentItem
+      .where(:content_id.nin => ["", nil])
+      .where(:schema_name.not => /^placeholder/)
+  end
+
+  def remaining_routes
+    Set.new(routes.keys) - checked_routes
+  end
+
+  def get_route(path)
+    begin
+      route = routes.fetch(path.to_sym)
+      checked_routes.add?(path.to_sym)
+      route
+    rescue KeyError
+      errors[path] << "Path (#{path}) was not found!"
+      nil
+    end
+  end
+
+  def find_content_item_by_route(path)
+    ContentItem.where("routes.path" => path).first
+  end
+
+  def find_content_item_by_redirect(path)
+    ContentItem.where("redirects.path" => path).first
+  end
+
+  def find_content_item(path)
+    # We cannot use an 'or' query here as it seems to be slower than doing this
+    # irb(main):019:0> time { ContentItem.where(routes: { "$elemMatch" => { path: path } }).first }
+    #   0.000000   0.000000   0.000000 (  0.006512)
+    # irb(main):020:0> time { ContentItem.where(redirects: { "$elemMatch" => { path: path } }).first }
+    #   0.010000   0.000000   0.010000 (  0.004688)
+    # irb(main):021:0> time { ContentItem.or({ redirects: { "$elemMatch" => { path: path } } }, { routes: { "$elemMatch" => { path: path } } }).first }
+    #   0.000000   0.000000   0.000000 ( 11.314430)
+    find_content_item_by_route(path) || find_content_item_by_redirect(path)
```

Comment:
is it a concern of ours (ie within scope of this) where there could be either multiples of these or both of these? - just thinking that we only do something on the first thing we found whereas there is capacity for this to be multiple.

Also I'd guess (but don't know) that we know from the router data whether this is a redirect or not so might be able to limit the scope of this or based on the data?